### PR TITLE
EXSWHTEC-262 - Introduce build dependencies for atomic arithmetic operations

### PR DESCRIPTION
**Pull request by mirza-halilcevic**

### CODE REVIEW

File: catch/unit/atomics/arithmetic_common.hh [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+  if constexpr (use_shared_mem) {
+    if (tid < width) {
+      const auto target = PitchedOffset(mem, pitch, tid);
+      *target = *PitchedOffset(global_mem, pitch, tid);
+    };
+    __syncthreads();
+  }
+
+  old_vals[tid] =
+      PerformAtomicOperation<TestType, operation>(PitchedOffset(mem, pitch, tid % width));
```
Comment (b-sumner) [inline catch/unit/atomics/arithmetic_common.hh]: This test is fine, but I was hoping to see some (or a lot) of non-atomic traffic produced by other wavefronts or blocks competing with these atomic operations in the memory system, including hitting in the same cache line where the atomic variable resides.

Comment (mirza-halilcevic) [reply catch/unit/atomics/arithmetic_common.hh]: As per our discussion, such test cases, as well as other more sophisticated ones including different memory orderings, will be implemented later on as part of the tests for the built-in atomic functions. For now, these tests are meant to be a solid foundation to rely on.

Comment (b-sumner) [reply catch/unit/atomics/arithmetic_common.hh]: OK, thanks for the reminder.  I agree this is a good foundation.

**Comment by mangupta** [comment]
@mirza-halilcevic : Could you please look into splitting this PR into smaller ones so that it can be reviewed easily?

File: catch/unit/atomics/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+# THE SOFTWARE.
+
+set(TEST_SRC
+)
+
+if(HIP_PLATFORM MATCHES "nvidia")
+    hip_add_exe_to_target(NAME AtomicsTest
+                        TEST_SRC ${TEST_SRC}
+                        TEST_TARGET_NAME build_tests
+                        LINKER_LIBS "nvrtc -rdc=true -gencode arch=compute_60,code=sm_60 -gencode arch=compute_70,code=sm_70 -gencode arch=compute_80,code=sm_80")
```
Comment (scchan) [inline catch/unit/atomics/CMakeLists.txt]: Any reason why this requires the RDC option?  Why this isn't needed for the amd platform?

File: catch/include/resource_guards.hh [hunk 3]
```diff
@@ -219,7 +263,28 @@ class StreamGuard {
...
+
+  StreamGuard(StreamGuard&& o)
+      : stream_type_{o.stream_type_}, flags_{o.flags_}, priority_{o.priority_}, stream_{o.stream_} {
+    o.stream_type_ = Streams::nullstream;
+    o.flags_ = 0u;
+    o.priority_ = 0;
+    o.stream_ = nullptr;
+  }
+
+  StreamGuard& operator=(StreamGuard&& o) {
```
Comment (scchan) [inline catch/include/resource_guards.hh:272]: This needs to check whether it's assigning to itself.

File: catch/include/resource_guards.hh [hunk 4]
```diff
@@ -219,7 +263,28 @@ class StreamGuard {
...
 
   StreamGuard(const StreamGuard&) = delete;
-  StreamGuard(StreamGuard&&) = delete;
+
+  StreamGuard(StreamGuard&& o)
+      : stream_type_{o.stream_type_}, flags_{o.flags_}, priority_{o.priority_}, stream_{o.stream_} {
+    o.stream_type_ = Streams::nullstream;
+    o.flags_ = 0u;
+    o.priority_ = 0;
+    o.stream_ = nullptr;
```
Comment (scchan) [inline catch/include/resource_guards.hh]: This entire block could simply be replacement by using the move assignment:
*this = std::move(o);

File: catch/include/resource_guards.hh [hunk 5]
```diff
@@ -55,15 +75,36 @@ template <typename T> class LinearAllocGuard {
...
-  LinearAllocGuard(LinearAllocGuard&&) = delete;
+
+  LinearAllocGuard(LinearAllocGuard&& o)
+      : allocation_type_{o.allocation_type_}, ptr_{o.ptr_}, host_ptr_{o.host_ptr_} {
+    o.allocation_type_ = LinearAllocs::noAlloc;
+    o.ptr_ = nullptr;
+    o.host_ptr_ = nullptr;
+  }
+
+  LinearAllocGuard& operator=(LinearAllocGuard&& o) {
```
Comment (scchan) [inline catch/include/resource_guards.hh:88]: This needs to check whether it's assigning to itself.

File: catch/include/resource_guards.hh [hunk 6]
```diff
@@ -55,15 +75,36 @@ template <typename T> class LinearAllocGuard {
...
   }
 
   LinearAllocGuard(const LinearAllocGuard&) = delete;
-  LinearAllocGuard(LinearAllocGuard&&) = delete;
+
+  LinearAllocGuard(LinearAllocGuard&& o)
+      : allocation_type_{o.allocation_type_}, ptr_{o.ptr_}, host_ptr_{o.host_ptr_} {
+    o.allocation_type_ = LinearAllocs::noAlloc;
+    o.ptr_ = nullptr;
+    o.host_ptr_ = nullptr;
```
Comment (scchan) [inline catch/include/resource_guards.hh]: This entire block could simply be replacement by using the move assignment:
*this = std::move(o);

File: catch/include/cmd_options.hh [hunk 7]
```diff
@@ -23,11 +23,11 @@ THE SOFTWARE.
 #pragma once
 
 struct CmdOptions {
-  int iterations = 1000;
+  int iterations = 10;
```
Comment (scchan) [inline catch/include/cmd_options.hh]: This change requires some explanation and may need proof of a consensus from the HIP leads.

**Review by scchan — CHANGES_REQUESTED** [review]

File: catch/include/cmd_options.hh (hunk 7 above)
Comment (scchan) [reply catch/include/cmd_options.hh]: @mangupta @gargrahul please review this change to the number of iterations.

File: catch/include/resource_guards.hh [hunk 8]
```diff
@@ -29,10 +29,30 @@ enum class LinearAllocs {
...
+    case LinearAllocs::mallocAndRegister:
+      return "registered";
+    case LinearAllocs::hipHostMalloc:
+      return "host pinned";
+    case LinearAllocs::hipMalloc:
+      return "device malloc";
+    case LinearAllocs::hipMallocManaged:
+      return "managed";
+    default:
+      return "unknown alloc type";
```
Comment (b-sumner) [inline catch/include/resource_guards.hh:48]: I'm not sure these strings are informative enough and/or translate easily to API usage.

File: catch/include/resource_guards.hh [hunk 9]
```diff
@@ -55,15 +75,36 @@ template <typename T> class LinearAllocGuard {
...
+  LinearAllocGuard(LinearAllocGuard&& o)
+      : allocation_type_{o.allocation_type_}, ptr_{o.ptr_}, host_ptr_{o.host_ptr_} {
+    o.allocation_type_ = LinearAllocs::noAlloc;
+    o.ptr_ = nullptr;
+    o.host_ptr_ = nullptr;
+  }
+
+  LinearAllocGuard& operator=(LinearAllocGuard&& o) {
+    allocation_type_ = o.allocation_type_;
+    ptr_ = o.ptr_;
```
Comment (scchan) [inline catch/include/resource_guards.hh]: The `allocation_type` and `ptr_` need to be checked and freed if needed.

File: catch/unit/atomics/CMakeLists.txt (hunk 2 above)
Comment (mirza-halilcevic) [reply catch/unit/atomics/CMakeLists.txt]: Seems to just be mistakenly left in. It is not needed.

File: catch/include/resource_guards.hh (hunk 3 above)
Comment (mirza-halilcevic) [reply catch/include/resource_guards.hh:272]: Done.

File: catch/include/resource_guards.hh (hunk 4 above)
Comment (mirza-halilcevic) [reply catch/include/resource_guards.hh]: Done.

File: catch/include/resource_guards.hh (hunk 5 above)
Comment (mirza-halilcevic) [reply catch/include/resource_guards.hh:88]: Done.

File: catch/include/resource_guards.hh (hunk 6 above)
Comment (mirza-halilcevic) [reply catch/include/resource_guards.hh]: Done.

File: catch/include/resource_guards.hh (hunk 9 above)
Comment (mirza-halilcevic) [reply catch/include/resource_guards.hh]: Done.

File: catch/include/resource_guards.hh (hunk 8 above)
Comment (mirza-halilcevic) [reply catch/include/resource_guards.hh:48]: Agree. I changed them so they reference the exact function used.

**Review by scchan — APPROVED** [review]
LGTM thanks

**Comment by b-sumner** [comment]
LGTM too.

**Comment by rakesroy** [comment]
PR has been merged into develop branch via commit 2eb17f0696c3ff9f45d15d1d74a3c40f84fbeb18.

**Comment by rakesroy** [comment]
PR has been merged into develop branch via commit b6180a3354465f18d5147ed588b25ca7a2147830.

**Pull request closed without merging**